packages/oauth/oauth-provider/src/lib/util/zod-error.ts

```diff
@@ -2,12 +2,31 @@ import { ZodError } from 'zod'
 
 export function extractZodErrorMessage(err: unknown): string | undefined {
   if (err instanceof ZodError) {
-    const issue = err.issues[0]
-    if (issue?.path.length) {
-      // "part" will typically be "body" or "query"
-      const [part, ...path] = issue.path
-      return `Validation of "${path.join('.')}" ${part} parameter failed: ${issue.message}`
+    for (const issue of err.issues) {
+      if (issue.path.length) {
+        // "part" will typically be "body" or "query"
+        const [part, ...path] = issue.path
+        const title = path.length
+          ? `Validation of  "${path.join('.')}" ${part} parameter failed`
+          : `Invalid request ${part}`
+        const message = extractMessage(issue)
+        return message ? `${title}: ${message}` : title
+      }
     }
+
+    const message = extractMessage(err)
+    if (message) return `Invalid request: ${message}`
+  }
+
+  return undefined
+}
+
+function extractMessage({ message }: { message: string }): string | undefined {
+  // message serialization defaults to JSON.stringify(err.issues)
+  if (!message.startsWith('[') && !message.endsWith(']')) {
```

Comment:
If the purpose of the logic is to detect a serialized JSON array, I believe you'd want to use `||` here.

Reply:
this code was updated to rely on the issue messages instead of error messages, removing the need for this heuristic.